### PR TITLE
Batch embedding load

### LINKED ISSUE #2339: [Feature]: Cache Embedding data for parsing on model change.
### Feature description

It's a known issue that I have what may be considered "too many of all the things," but re-parsing Embeddings is impacting model switching for me. I haven't delved into it, I am more putting this Issue here so I can look in to it and have an Issue to post to/link to.

### Version Platform Description

_No response_

### PR DESCRIPTION
**Pull request by midcoastal**
## Description

Optimize performance with Embedding Loading

## Notes

Patch methods in `diffusers.loaders` to allow bulk TI loading.
- Implement selectable exception suppression to skip bad Embeds.
- Track Embed Filenames with Embed Shapes/Tokens to track loaded.
Refactor `load_diffusers_embedding` and `load_from_dir` to suit.
- As with `diffusers.loaders`, use Exceptions for bad Embed skipping.
- Since Embeds may be multi-vector, track each vector token and shape.
- Sanity-check that MV Tokens aren't pre-existing.
- Track Diff Model Type and Embed Type each.
`list_embeddings` now omits exts unsupported by DIFFUSERS
Lots of `debug` reports.  Perhaps too many, but who's to say?

What is going on here?
A massive performance hit was noticed in calling pipe.load_textual_inversion()
for each TI, as each call causes the model to rebuild it's Trie, which
is incredibly expensive.

On this developers test system, the following timings were observed:
- Loading 0 Embeddings, WebUI thinks it takes ~8s to "Load Embeddings"
- Loading ~3100 Embeddings, without the relevant changes, takes ~1220s
- Loading ~3300 Embeddings, WITH the relevant changes, takes ~51s

closes #2339

### CODE REVIEW

**Comment by midcoastal** [comment]
As an aside, during testing I noticed that `DIFFUSERS` handles `BREAK` differently (and breaks).

Example using an SD1.5 model:

`wood elf rogue, BREAK (white_background:1.4), illustration, absurdres, rpg dnd fantasy, character_portrait, solo, full_body, BREAK traveller rpg`
image

`wood elf rogue, BREAK (white_background:1.4), illustration, absurdres, rpg dnd fantasy, character_portrait, solo, full_body, traveller rpg`
image

Examples using SD-XL model (same prompts):

image

image

In theory, the added `BREAK`s shouldn't change the image "much", or at least, "drastically."  For example, here is the same prompt without *any* `BREAK`s:

image

Pretty much the same as the one with `BREAK`, but still *very* different from the one with 2.

Am I missing something?  This sent me down a *very* annoying rabbit hole...  :-(

**Comment by vladmandic** [comment]
lets have a conversation about BREAK in some separate thread.

regarding this PR, my issue is that it bundles 3 unrelated items in a single PR.
- changes to `installer.py` - why is that there?
- changes to `modelloader.py` - there is some enhanced functionality which is fine, but it also undoes a lot of changes i did. why rename params back from `path` to `dir`? i did that so it passes lint checks - `dir` is python built-in and i don't want to have chance of any variable overriding built-ins, regardless of how safe it is. note that sdnext comes with `ruff` and `pylint` rules - so please run those two on your code.
- changes to how actual textual inversions are loaded. once this is separated from other two changes, i will test it.

**Comment by midcoastal** [comment]
Thank you for the feedback, let's do that.

As for the changes:

1. Changes in `installer.py` are with regards to patching the logger.  I had noticed in Extra Networks there was a check against an environment variable for displaying additional debugging.  I needed something similar while I was developing, but I didn't want to have to put `if debug:` everywhere.  Further, I wanted to be able to add a prefix to the messages, so I wasn't typing the same text over and over.  
  
   And lastly, for instances where I was just sending to `info`, I didn't want to have to keep typing `debug.info(...` and instead it would be helpful to just type `debug(`.  I figured, since I didn't change any existing signatures, and just added helpful/convenient functionality that I was actively using in my development, it wouldn't be an issue.

  In debugging and research, I actually added a ton of such debug calls to the code (I think 13 different environment variables, several hundred calls), I just didn't check them in, as they weren't directly relevant to this PR aside from research.

2. I forked from `master` way prior to the latest release, and was using terminology that was there/I was used to. When I merged master back in, I wasn't aware that the name changes were intentional.  I would be happy to fix the naming, in other locations as well.  Not an issue.

   As for `ruff` and `pylint`, do those get run as part of PR checks?  It looks like it passed (and wouldn't have notified me of issues). But I will take this to heart, and be more mindful of variable names.  I'l get on that.

3. Ok, so:  Push strictly changes, no debug entries (or, if so, in a different manner), and no changed to the directory/file listings.  If I were interested in distilling the directory/file listing functions, or adding functionality to aid in additional permanent debug statements, I should do that in a separate PR, even if I am using those improvements/added them during the development of something else?  Yes?

   That is going to be a tough practice for me to get into, as I often discover things that need to be fixed while working on other things, or come up with something that would be helpful both in the moment, as well as the future, and begin immediately using it.  I am not sure how to approach that kind of a situation if it's desired to have each portion in a new PR, especially if the code in each relies on something in another.

   What is the best practice, there?

**Comment by midcoastal** [comment]
Also, I should note:  There are a few just cleanup/polishing changes that are inbound anyway.  So, pause on approving this PR anyway.  Should be quick.

**Comment by vladmandic** [comment]
yup, everything you said holds. and only real best practice is that each pr should be self contained.
two possible (different) scenarios: 
- if you address multiple things while working on a pr, you can make a branch out of your fork and and then in one branch revert changes that are not related to pr and another branch use to submit second pr with the rest of the changes.
- if you have chained requirements, you can mentione that explicitly - normally i'm pretty quick addressing those. e.g., submit one pr for "i will need those changes as prerequisite for xxx" and continue working on xxx (hoping) that that first pr will make it in by the time your ready to submit actual main pr.

**Comment by midcoastal** [comment]
Ok, I'll keep that in mind.

So, with that, what do you suggest here?  I can totally back out either of the others without too much issue.  They are both pretty much just "nice to have." I need to A/B test the modelloader stuff anyway.  What I ended up with "felt" better, and in testing addressed some separate concerns I had, but "performance wise" it is within seconds of the other (as scale), so, I don't think too big an issue for the PR at hand.

Then submit PR for review for the others?

**Comment by vladmandic** [comment]
- make a branch out of your fork so you don't lose changes
- back out installer and modelloader changes for this pr.
- re-target pr to target dev branch
then you can decide later if you want to submit those two other changes separately (and rebase to use current dev due to param names, etc.)

**Comment by midcoastal** [comment]
I am going to do just that.  Thank you for the feedback.

On the `dir` thing: it is common in the code so far to use `dir` as the short form for `directory`.  In fact, I thought to mention it when I went in to patch it all to `path` and noted how often I kept hitting `model_dir`, etc. I can understand not wanting to clobber a python built-in, even in the function-level local scope, and I will do what I can to work within that.  That said, I wold prefer, then, to use `directory` in full, as `path` is ambiguous, I feel.  Would that be acceptable?

**Comment by midcoastal** [comment]
Or `dir_path`.  I just did a quick search, and I am impressed that you did clean all of the `dir`s, save for the hyphenated or snake-cased ones. You did miss ONE file, though: `modules/ui_tempdir.py`  ;-)

**Comment by vladmandic** [comment]
i'm ok with `directory`, `folder`, `dir_path` - pretty much all of them are clean enough, just don't want to add exceptions to allow clobber. one in `modules/ui_tempdir` is left because `gradio` refers to it by name (its a function that hijacks gradio's internal handling of temp files).

**Comment by vladmandic** [comment]
converting to draft until changes can be isolated to ti loading only.

**Comment by midcoastal** [comment]
Thank you.

**Comment by midcoastal** [comment]
@vladmandic - Ready for review/test/merge.

**Comment by midcoastal** [comment]
@vladmandic - Of note: When merging these two together into `dev` there is a line that needs to be removed.  Line 13 in `textual_inversion.py`.  I tried tracking the change, but no dice.  It's manual.

**Comment by midcoastal** [comment]
@vladmandic I just wanted to round back to this. I have been trying to keep it up to date with dev, and wanted to see if I should continue maintaining these (two) pull requests.

If so, I think I will just completely re-base them, to make the commit history cleaner, but I'm not going to do that unless you are still interested in considering them.

Please advise.

**Comment by midcoastal** [comment]
> converting to draft until changes can be isolated to ti loading only.

@vladmandic - Current changes have been isolated.  The changes in `patches` are used in patching the `diffusers` code.

I have updated this to work with the latest version of `diffusers` (they changed folder structure), and have been using this locally without issue for a bit now.

Do you have any other requests/thoughts?

**Comment by vladmandic** [comment]
i don't have open requests at the moment, i just need to dedicate time to review and test this.

**Comment by midcoastal** [comment]
> i don't have open requests at the moment, i just need to dedicate time to review and test this.

@vladmandic - Would it be beneficial for me to provide some more detailed explanation, in chunks, (either in comments in this PR, or in the code proper) as to what exactly I am doing, my intent, and why, and why it is different than the base methods?

**Comment by vladmandic** [comment]
i don't think thats needed, code is quite readable as-is

**Comment by midcoastal** [comment]
@vladmandic - Latest version of `diffusers` does add an additional logic check in this method.  It is highly unlikely that `SD.Next` will trigger the logic check, but I am going to modify to suit.  Further, I am guessing that a hangup here is the fact that I am using list comprehension/generators fairly heavily, would it be beneficial, since I am going to be in there, to explode those comprehensions?

**Comment by vladmandic** [comment]
this has been in my backlog and i didn't want to move forward due to huge number of other changes in the last release - it was just too much to do it all at once.

moving forward, yes, this can be reviewed and merged.
i don't have structural issues with how pr is implemented, but due to sheer size, i will need to test it.

let me know when its up-to-date as per your last comment.

**Comment by midcoastal** [comment]
Should be good to go for current `dev`

**Comment by midcoastal** [comment]
For reference:

```
Dev No Network:
	Average Startup time (1 samples): 30.48
	Average VAEs time (1 samples): 7.59
	Average Models time (1 samples): 3.92
	Average Extensions time (1 samples): 6.41
	Average Networks time (1 samples): 1.44
	Average Model time (30 samples): 5.06
	Average Embeddings time (28 samples): 0.56
	Average time loading 0 Embeddings (12 samples): 0.63 (47.8723 E/s)
	Average time loading 30 Embeddings (16 samples): 0.51 (58.4653 E/s)

Dev Over Network:
	Average Startup time (1 samples): 435.45
	Average VAEs time (1 samples): 22.01
	Average Models time (1 samples): 20.05
	Average Extensions time (1 samples): 201.73
	Average Networks time (1 samples): 188.67
	Average Model time (16 samples): 9.51
	Average Embeddings time (11 samples): 586.88
	Average time loading 4302 Embeddings (5 samples): 1271.74 (3.5448 E/s)
	Average time loading 57 Embeddings (6 samples): 16.16 (278.9892 E/s)

Branch No Network:
	Average Startup time (5 samples): 10.69
	Average VAEs time (5 samples): 1.01
	Average Models time (5 samples): 2.89
	Average Extensions time (5 samples): 3.85
	Average Networks time (5 samples): 0.9
	Average Model time (59 samples): 4.55
	Average Embeddings time (53 samples): 0.43
	Average time loading 0 Embeddings (22 samples): 0.58 (51.7241 E/s)
	Average time loading 30 Embeddings (31 samples): 0.32 (93.7500 E/s)
	
Branch Over Network:
	Average Startup time (1 samples): 410.52
	Average VAEs time (2 samples): 21.66
	Average Models time (2 samples): 19.52
	Average Extensions time (2 samples): 190.64
	Average Networks time (1 samples): 183.14
	Average Model time (20 samples): 10.9
	Average Embeddings time (19 samples): 32.92
	Average time loading 57 Embeddings (7 samples): 16.96 (267.1776 E/s)
	Average time loading 65 Embeddings (1 samples): 40.99 (110.2464 E/s)
	Average time loading 4301 Embeddings (11 samples): 39.38 (114.7669 E/s)
```

File: modules/textual_inversion/textual_inversion.py [hunk 1]
```diff
@@ -127,83 +137,163 @@ def get_expected_shape(self):
...
+                            filename = os.path.basename(path)
+                        fn, ext = os.path.splitext(filename)
+                        name = os.path.basename(fn)
+                        embedding = Embedding(vec=None, name=name, filename=path)
+                        try:
+                            ext  = ext.upper()
+                            _, _ext = os.path.splitext(path)
+                            _ext = _ext.upper()
+                            if ext != _ext:
+                                raise ValueError(f'filename and path extensions do not match: `{ext}` != `{_ext}`')
```
Comment (vladmandic) [inline modules/textual_inversion/textual_inversion.py:185]: why is this part needed?

**Comment by vladmandic** [comment]
any updates on this pr? i'd like to try it out in this release cycle...

**Comment by midcoastal** [comment]
@vladmandic - Oh I would love to, but I seem to have run in to a bit of a problem...

Something strange happen on `dev` where files are looking like they are being completely rewritten with the same data...  This is breaking `diff`s and `merge`s as it is impossible to see what is actually an incoming change and what is not.  It is also going to cause all sorts of havoc with `git blame` and file histories...

Care to explain whats going on?

**Comment by vladmandic** [comment]
global crlf -> lf cleanup that was done 1h ago.

**Comment by midcoastal** [comment]
Um...  What? I mean, its your repo, but in absolutely every `git` repo i have *ever* been involved with, that was one of the fundamental things we didn't do... When i had just started, professionally, that actually got me removed from a project... 

Your call. But this seems like a nightmare.

Enforce line endings in changes. That way the blames, commits, and PR/Issues remain linked correctly.

**Comment by midcoastal** [comment]
I actually have a `git` action I use just for this purpos, of you want me to submit it...

**Comment by midcoastal** [comment]
Just being honest, but I have no interest in resolving these merge conflicts.

**Comment by vladmandic** [comment]
i'm aware of the git action for that, as well as the git config params.
problem was a lot of legacy and 3rd party code that was imported into repo.
this may be painful, but a one time cleanup was my call. now i can enforce it on git level moving forward.

and regarding merge conflicts - if you do the same conversion of your code, then merge check would basically just go away.

**Comment by midcoastal** [comment]
@vladmandic - I do believe it is in good shape.

**Comment by midcoastal** [comment]
@vladmandic - ugh, please hold.  some sort of merge conflict...

**Comment by midcoastal** [comment]
@vladmandic - provided this passes Ruff, it should be good to go now.

**Comment by vladmandic** [comment]
I have some time this week to go over it.

**Comment by midcoastal** [comment]
I just double-checked, and the timings all look consistent.  There is a slight uptick in overall timings, but that is consistent with a slight uptick with raw `dev`, and the difference is uniform, so, still worth the go.

**Comment by vladmandic** [comment]
this PR is now merged. i just did couple of minor cleanups after the merge.

**Comment by midcoastal** [comment]
@vladmandic - I couldn't help but notice this has been pretty much pulled out.  Mind if I ask what happen?  My load times have gone from ~30 seconds, back up to ~300 seconds.  Did I miss something?

**Comment by midcoastal** [comment]
I will also be sure to heavily document/comment my code from now on, to hopefully better communicate why I am doing particular things, as I had offered earlier.

**Comment by vladmandic** [comment]
the problem was that this pr basically broke embeddings completely for many use-cases.
it was also corrupting sdxl encoder and resulting in cuda dsa errors. entire discussion history is on discord and ai-casanova jumped in with a quick rewrite.
its not about documenting/commenting, its about a) simplifying, b) testing all different scenarios - sdnext is cross-platform and cross-model app.

**Comment by midcoastal** [comment]
That's curious. I will check the discord. My changes didn't change any behavior beyond batching the loads. I will be interested to see how it breaks. I was using it with no discernible problems.

I'll have a look.

**Comment by midcoastal** [comment]
I haven't looked yet, but IIRC, there is one `if` that seemed sus to me that I am guessing caused the corruption. If that's the case, then yeah, better documentation on my part probably would have been a good idea. For ease of maintenance down the road I am going to do better about that regardless.

**Comment by vladmandic** [comment]
referencing multiple loaded sdxl embeddings caused torch runtime errors due to corruption in text encoder.

**Comment by midcoastal** [comment]
> referencing multiple loaded sdxl embeddings caused torch runtime errors due to corruption in text encoder.

Thank you for the hint.

**Comment by midcoastal** [comment]
> referencing multiple loaded sdxl embeddings caused torch runtime errors due to corruption in text encoder.

Yeah, haven't looked yet, but I know exactly what happen.
